google/cloud/documentai_toolbox/wrappers/page.py

```diff
@@ -411,9 +603,73 @@ def __post_init__(self):
         self.form_fields = _get_form_fields(
             form_fields=self.documentai_page.form_fields, text=self.text
         )
-        self.lines = _get_lines(lines=self.documentai_page.lines, text=self.text)
+
+        self.tokens = _get_tokens(tokens=self.documentai_page.tokens, text=self.text)
+        self.lines = _get_lines(
+            lines=self.documentai_page.lines, text=self.text, tokens=self.tokens
+        )
         self.paragraphs = _get_paragraphs(
-            paragraphs=self.documentai_page.paragraphs, text=self.text
+            paragraphs=self.documentai_page.paragraphs, text=self.text, lines=self.lines
+        )
+        self.blocks = _get_blocks(
+            blocks=self.documentai_page.blocks,
+            text=self.text,
+            paragraphs=self.paragraphs,
         )
-        self.blocks = _get_blocks(blocks=self.documentai_page.blocks, text=self.text)
         self.tables = tables
+
+    def to_hocr(self, filename):
+        r"""Exports a string hOCR version of the documentai.Document.Page.
+
+        Args:
+            filename (str):
+                Required. The filename of the original file.
+
+        Returns:
+            str:
+                A string hOCR version of the documentai.Document.Page.
+        """
+        f = ""
+        dimensions = self.documentai_page.dimension
+        pidx = self.documentai_page.page_number
+        page_bounding_box = _get_hocr_bounding_box(
+            element_with_layout=(self.documentai_page), dimensions=(dimensions)
+        )
+        f += f"<div class='ocr_page' lang='unknown' title='image \"{filename}\";{page_bounding_box}'>\n"
+        for bidx, block in enumerate(self.blocks):
```

Comment:
Instead of nested for-loops, is it feasible to have an `to_hocr` method in the wrapper classes Page, Block, Line, Token and call it recursively?  (This may be an abstract method in a parent base class for all wrapper classes that support `to_hocr`.) 

So the method body becomes something similar to:

```
class _HocrBase:
  @abstractmethod
  @property
  def children(self) -> List[Any]:
    ...

  def to_hocr(self, filename):  # ideally we don't have to ask for filename
    result = "..."  # page-level info
    for i, child in self.children:
      result += child.to_hocr(filename)
    result += "</span>"  # page-level.
```

I think this can be considered a follow-up refactoring work and does not need to be done in this PR.

Reply:
i think for the purpose of releasing this and letting the customer be able to use this, we should skip this and add it to backlog for refactoring

Reply:
Agreed.  Let's add an issue of low-ish priority to track this.  This kind of refactoring makes things easier to test locally (we can test each level separately), and when things break it will be much easier to pinpoint where things break since the tests would be smaller.